### PR TITLE
feat(modal): Overlay information component

### PR DESCRIPTION
**Pull request by Leotheluck**
A Modal is used to provide some information to users that needs a response from them.

While a Modal is triggered, there is no interaction possible on the page that is overlaid so users have to dismiss or click on an action button to proceed further.

Modals are used in different cases:

    alerting users about something that requires their agreement
    confirming a user decision
    notifying the user of an important information

There are four different sub-types of standard Modals to inform users that a problem requires immediate response from them so the process can continue, depending on the usage:

    Error: reserved for errors, malfunctions, as well as critical issues
    Warning: reserved for Modals that need the user attention and acknowledgment but might not cause errors
    Information: provides information to users in context
    Success: reserved to provide a static persistent success information

They can be dismissible or not.

Usage:
`<osds-modal
    color="info"
    headline="On Vous Héberge ?"
    dismissible="true"
  >
    <osds-text color="text">
      OVHcloud, anciennement OVH, est une entreprise française. Elle pratique initialement de l'hébergement de serveur, et est un fournisseur d'accès à Internet (FAI), puis opérateur de télécommunications pour les entreprises. Elle se développe, à la fin des années 2010, dans le cloud computing (informatique en nuage).
    </osds-text>
    <osds-button slot="actions" color="default" href="https://www.ovh.com/fr/" target="_blank">En savoir plus</osds-button>
    <osds-button slot="actions" color="primary" href="https://www.ovh.com/fr/" target="_blank">On m'héberge</osds-button>
  </osds-modal>`

### CODE REVIEW

File: packages/stencil/components/modal/src/components/osds-modal/osds-modal.scss [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    transform: scale(0.8);
+  }
+  to {
+    opacity: 1;
+    transform: scale(1);
+  }
+}
+
+:host {
+  display: flex;
```
Comment (dpellier) [inline packages/stencil/components/modal/src/components/osds-modal/osds-modal.scss:23]: probably not needed

File: packages/stencil/components/modal/src/components/osds-modal/osds-modal.tsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                  onClick: () => this.close(),
+              }}></osds-icon>
+            }
+          </div>
+
+          <div class="content">
+            {headline && headline.length > 0 &&
+              <div class="headline">
+                <osds-text
+                  {...{
```
Comment (dpellier) [inline packages/stencil/components/modal/src/components/osds-modal/osds-modal.tsx:97]: same

File: packages/stencil/components/modal/src/components/osds-modal/styles/osds-modal.size.scss [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        height: calc(var(--ods-size-09) + var(--ods-size-04));
+        padding: 0 calc(var(--ods-size-04) / 2);
+    }
+
+    :host .wrapper .content {
+        width: 100%;
+        padding: var(--ods-size-07);
+    }
+
+    :host .wrapper .content .headline {
```
Comment (dpellier) [inline packages/stencil/components/modal/src/components/osds-modal/styles/osds-modal.size.scss:45]: same as above regarding wrapped syntax

File: packages/stencil/components/modal/src/global.dev.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+// it always has to include './components' and './global.ts'
+// ###
+
+import './components';
+import './global';
+import { OdsLogger } from '@ovhcloud/ods-core';
+
+import '@ovhcloud/ods-stencil/components/icon';
+import '@ovhcloud/ods-stencil/components/text';
+import '@ovhcloud/ods-stencil/components/button';
```
Comment (dpellier) [inline packages/stencil/components/modal/src/global.dev.ts:12]: rebase and use new ones (careful to add dependencies accordingly)

**Review by dpellier — CHANGES_REQUESTED** [review]
`inert` is not removed from body after closing the modal (tried on index.html)

The way it is implemented, how will it behave if you put multiple osds-modal accross an app?

File: packages/libraries/core/src/components/modal/ods-modal-controller.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ * it contains all the glue between framework implementation and the third party service.
+ */
+export class OdsModalController extends OdsComponentController<OdsModal> {
+  private readonly logger = new OdsLogger('OdsInputController');
+
+  constructor(component: OdsModal) {
+    super(component);
+  }
+
+  close() {
```
Comment (aesteves60) [inline packages/libraries/core/src/components/modal/ods-modal-controller.ts:16]: need spec test for the controller

File: packages/stencil/components/modal/src/docs/osds-modal/usage.mdx [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import GenericStyle from '@ovhcloud/ods-core/docs/generic-style.mdx';
+
+<GenericStyle />
+
+## Usage
+
+### Default
+<osds-modal>Modal</osds-modal>
```
Comment (aesteves60) [inline packages/stencil/components/modal/src/docs/osds-modal/usage.mdx:8]: Can you add a button trigger ?

File: packages/stencil/components/modal/src/components/osds-modal/osds-modal.tsx [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    } = this;
+
+    return (
+      <Host {...{ masked }}>
+        <div class="backdrop"></div>
+
+        <div class="wrapper">
+          <div class="header">
+            { dismissible &&
+              <osds-icon
```
Comment (aesteves60) [inline packages/stencil/components/modal/src/components/osds-modal/osds-modal.tsx:82]: We can't focus this "button" with navigation key. I think for this you need to wrap the icon with a circle button

**Review by aesteves60 — CHANGES_REQUESTED** [review]

**Comment by ovh-cds** [comment]
CDS Report build#1183.0 ✘  
* 
  * build ✘

**Review by dpellier — CHANGES_REQUESTED** [review]
close button is bigger than the header, see
Capture d’écran 2023-08-17 à 11 52 24
(more easy to see on click hold)

**Review by feoche — CHANGES_REQUESTED** [review]
- [x] Backdrop overlay should stay the same color, as --ods-color-primary-500 with 75% opacity
- [x] There are errors on events on Storybook story
- [x] The #root is `inert` in the story, so it can't be interacted with
- [x] Closing icon should be in size `sm`
- [x] Content padding should be `padding: var(--ods-size-06) var(--ods-size-08) var(--ods-size-08)`
- [x] Modal title margin-bottom should be `var(--ods-size-04)`
- [x] Modal actions margin-top should be `var(--ods-size-07)`
- [x] Wrapper width should behave like this, without media queries `max-width: clamp(36rem, 70%, 70%); min-width: 50%;`

**Review by feoche — CHANGES_REQUESTED** [review]
- [ ] Clicking on Storybook buttons causes errors `args.onOpenModal is not a function`
- [x] Main slot should have `overflow-y: auto`

**Comment by ovh-cds** [comment]
CDS Report build#1224.0 ✘  
* 
  * build ✘

File: packages/libraries/core/src/components/public-api.ts [hunk 8]
```diff
@@ -28,3 +28,5 @@ export * from './ods-component-generic-methods';
 export * from './ods-component-methods';
 export * from './ods-focus-change-event-detail';
 export * from './ods-string-attributes';
+export * from './form-field/public-api';
```
Comment (dpellier) [inline packages/libraries/core/src/components/public-api.ts:31]: form-field has been moved

File: packages/stencil/components/modal/src/components/osds-modal/osds-modal.tsx [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+ * @slot (unnamed) - Modal content
+ */
+@Component({
+  tag: 'osds-modal',
+  styleUrl: 'osds-modal.scss',
+  shadow: true
+})
+export class OsdsModal implements OdsModal<OdsStencilMethods<OdsModalMethods>, OdsStencilEvents<OdsModalEvents>> {
+  controller: OdsModalController = new OdsModalController(this);
+  @Element() el!: HTMLElement;
```
Comment (dpellier) [inline packages/stencil/components/modal/src/components/osds-modal/osds-modal.tsx:24]: probably not needed

**Review by dpellier — CHANGES_REQUESTED** [review]

**Comment by ovh-cds** [comment]
CDS Report build#1269.0 ✘  
* 
  * build ✘

**Review by dpellier — APPROVED** [review]

**Review by aesteves60 — APPROVED** [review]

**Review by feoche — APPROVED** [review]

**Pull request merged**